src/toolchain/toolchain.ts

```diff
@@ -181,12 +182,84 @@ export class SwiftToolchain {
+        if (process.platform !== "darwin") {
+            return [];
+        }
         const { stdout: xcodes } = await execFile("mdfind", [
             `kMDItemCFBundleIdentifier == 'com.apple.dt.Xcode'`,
         ]);
         return xcodes.trimEnd().split("\n");
     }
 
+    /**
+     * Reads the swiftly configuration file to find a list of installed toolchains.
+     *
+     * @returns an array of toolchain paths
+     */
+    public static async getSwiftlyToolchainInstalls(): Promise<string[]> {
+        try {
+            const swiftlyHomeDir: string | undefined = process.env["SWIFTLY_HOME_DIR"];
+            if (!swiftlyHomeDir) {
+                return [];
+            }
+            const swiftlyConfigRaw = await fs.readFile(
+                path.join(swiftlyHomeDir, "config.json"),
+                "utf-8"
+            );
+            const swiftlyConfig: unknown = JSON.parse(swiftlyConfigRaw);
+            if (
+                swiftlyConfig === undefined ||
+                swiftlyConfig === null ||
+                typeof swiftlyConfig !== "object" ||
+                !("installedToolchains" in swiftlyConfig)
+            ) {
+                return [];
+            }
+            const installedToolchains = swiftlyConfig.installedToolchains;
+            if (!Array.isArray(installedToolchains)) {
+                return [];
+            }
+            return installedToolchains
+                .filter((toolchain): toolchain is string => typeof toolchain === "string")
+                .map(toolchain => path.join(swiftlyHomeDir, "toolchains", toolchain));
+        } catch (error) {
+            throw new Error("Failed to retrieve Swiftly installations from disk.");
+        }
+    }
+
+    /**
+     * Checks common directories for available swift toolchain installations.
+     *
+     * @returns an array of toolchain paths
+     */
+    public static getToolchainInstalls(): Promise<string[]> {
+        return Promise.all([
+            this.findToolchainsIn("/Library/Developer/Toolchains/"),
+            this.findToolchainsIn(path.join(os.homedir(), "Library/Developer/Toolchains/")),
+        ]).then(results => results.flatMap(a => a));
+    }
+
+    /**
+     * Searches the given directory for any swift toolchain installations.
+     *
+     * @param directory the directory path to search in
+     * @returns an array of toolchain paths
+     */
+    public static async findToolchainsIn(directory: string): Promise<string[]> {
+        try {
+            return (await fs.readdir(directory, { withFileTypes: true }))
+                .filter(
+                    dirent =>
+                        dirent.name.startsWith("swift-") &&
+                        (dirent.isDirectory() || dirent.isSymbolicLink())
+                )
+                .map(dirent => path.join(dirent.path, dirent.name));
+        } catch {
+            // Assume that there are no installations here
+        }
+        return [];
```

Comment:
nit: move the return to the `catch` to make it more clear that `[]` is returned for the error condition

Reply:
done.